OvmfPkg/PlatformPei/PlatformId.c

```diff
@@ -0,0 +1,660 @@
+ */
+STATIC
+EFI_STATUS
+ConsumeAsciiByteStr (
+  IN  UINT8  *Buffer,
+  IN  UINT8  *BufferEnd,
+  OUT UINT8  **Start,
+  OUT UINT8  **End
+  )
+{
+  EFI_STATUS  Status;
+  UINT8       EvtSize;
+
+  Status = SafeOffset (Buffer, BufferEnd, sizeof (EvtSize), Start);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  EvtSize = *Buffer;
+  // The spec does not allow for a zero-length string since it must be NUL-
+  // terminated.
+  if (EvtSize == 0) {
+    DEBUG ((DEBUG_ERROR, DPREFIX "Expected a NUL-terminated string\n"));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Status = SafeOffset (*Start, BufferEnd, EvtSize, End);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  if (*(*End - 1) != 0) {
+    DEBUG ((DEBUG_ERROR, DPREFIX "Expected a NUL-terminated string\n"));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+ * Reads a first 32-bit integer as a size, then determines if Buffer + 4 + size
+ * exceeds BufferEnd.
+ *
+ * @param[in] Buffer  The buffer to read from.
+ * @param[in] BufferEnd  The end of the buffer.
+ * @param[out] Start  The start of the string.
+ * @param[out] End  The byte after the last character of the string.
+ *
+ * @retval EFI_SUCCESS  The string was read successfully.
+ * @retval EFI_BUFFER_TOO_SMALL  The string's declared size was too large for
+ *                               the buffer.
+*/
+STATIC
+EFI_STATUS
+ConsumeUint32Str (
+  IN  UINT8  *Buffer,
+  IN  UINT8  *BufferEnd,
+  OUT UINT8  **Start,
+  OUT UINT8  **End
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      EvtSize;
+
+  Status = ParseUint32 (Buffer, BufferEnd, &EvtSize, Start);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  return SafeOffset (*Start, BufferEnd, EvtSize, End);
+}
+
+/**
+ * Validates that the locator is valid.
+ *
+ * PREREQUISITE: Locator <= LocatorEnd
+ *
+ * @param[in] Locator  The locator to validate.
+ * @param[in] LocatorEnd  The end of the locator.
+ * @param[in] LocatorType  The type of the locator.
+ *
+ * @retval EFI_SUCCESS  The locator is valid.
+ * @retval EFI_BUFFER_TOO_SMALL  The locator is too small.
+ * @retval EFI_INVALID_PARAMETER  The locator is invalid.
+**/
+STATIC
+EFI_STATUS
+ValidateLocator (
+  IN UINT8   *Locator,
+  IN UINT8   *LocatorEnd,
+  IN UINT32  LocatorType
+  )
+{
+  UINTN  Size;
+
+  // No validation done for non-UEFI variable locators.
+  if (LocatorType != LOCATOR_TYPE_UEFI_VARIABLE) {
+    return EFI_SUCCESS;
+  }
+
+  Size = (UINTN)(LocatorEnd - Locator);
+  // The size needs to include the GUID and a UCS2 NUL terminated string, so it
+  // has to be at least 18 bytes, and be even.
+  if ((Size < sizeof (EFI_GUID) + 2) || ((Size & 1) != 0)) {
+    DEBUG ((DEBUG_ERROR, DPREFIX "UEFI variable locator improper size %d\n", Size));
+    return EFI_BUFFER_TOO_SMALL;
+  }
+
+  if (*((CHAR16 *)LocatorEnd - 1) != 0) {
+    DEBUG ((DEBUG_ERROR, DPREFIX "UEFI variable locator not UCS2 0-terminated\n"));
+    return EFI_INVALID_PARAMETER;
+  }

```

Comment:
This validation is very weak.

For example, a GUID shall only contain digital and a-f. But with this, many invalid GUID may pass the validation.

I think we need clarify the criteria in function header.

Reply:
The GUID is a 16 byte EFI_GUID, so any byte is valid. It's not the 36 character string. I think we can get rid of all validation if we do 1 event per file as you suggest.

Reply:
Yes, if we only do 1 event per file, that can definitely simplify the our work in EDKII.